enhancements/baremetal/baremetal-keepalived-unicast.md

```diff
@@ -0,0 +1,179 @@
+retrieving that node will need to be implemented to allow the control plane nodes to include the 
+bootstrap IP address in their Keepalived configuration. 
+
+##### Bootstrap node
+
+The bootstrap [Keepalived](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml) instance runs as static pod and the relevant [Keepalived configuration](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.conf.tmpl) rendered by the [baremetal-runtimecfg](https://github.com/openshift/baremetal-runtimecfg).
+
+The bootstrap Keepalived pod includes:
+- [Keepalived container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L35-#L93)
+- [TCP server container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L27#L34) that will be responsible to return the bootstrap IP address.
+- [keepalived-monitor container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L94-#L138) is responsible to render the Keepalived config file.
+
+The keepalived-monitor container in bootstrap node will first render Keepalived cfg file with an empty 'unicast_peer' section, as a result of that the `api-vip` will be set to the bootstrap node by Keepalived.
+
+At the next step, the bootstrap's keepalived-monitor container will repeatedly retrieve the control plane nodes details from the localhost:kube-apiserver and add them to 'unicast_peer' section.
+
+The bootstrap node owns the `api-vip' during bootstrap phase because it's configured with [higher](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.conf.tmpl#L13) VRRP priority than the control plane nodes [priority](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/master/00-master/baremetal/files/baremetal-keepalived-keepalived.yaml#L26).
+
+##### Control plane nodes
+
+The [Keepalived pod](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/common/baremetal/files/baremetal-keepalived.yaml) in control plane nodes already includes Keepalived and keepalived-monitor containers. 
+To support unicast mode, the keepalived-monitor should be updated to include the [bootstrap node IP address](https://github.com/openshift/baremetal-runtimecfg/blob/8242fd84d6c6d267d2d5551abf8e4f778b4f0615/pkg/monitor/dynkeepalived.go#L30) and [IP addresses of control plane nodes](https://github.com/openshift/baremetal-runtimecfg/blob/8242fd84d6c6d267d2d5551abf8e4f778b4f0615/pkg/config/node.go#L336-#L380).
+
```

Comment:
this should include information that the bootstrap ip is being fetching by `GET <apiVIP>:<ip discovery port>`

Is this discovery service running on all the control-plane nodes too, because in case i want to add a new control-plane machine, it will try to reach out to this IP:PORT for an ip and will probably fail to get all the information to join the cluster...

Reply:
That should be OK, the control-pane node repeatedly tries to GET bootstrap IP from <api-vip><discovery-port> in case no reply was received (after a timeout), Keepalived will be configured in this node only with the control-plane nodes as peers.

Reply:
Done